src/librustc_yk_sections/mir_cfg.rs

```diff
@@ -112,26 +121,59 @@ impl<'a, 'tcx, 'gcx> ConvCx<'a, 'tcx, 'gcx> {
     }
 }
 
-/// Converts and serialises the specified DefIds, returning an linkable ELF object.
+/// Writes TIR to file for the specified DefIds, possibly returning a linkable ELF object.
 pub fn generate_tir<'a, 'tcx, 'gcx>(
-    tcx: &'a TyCtxt<'a, 'tcx, 'gcx>, def_ids: &DefIdSet, exe_filename: PathBuf, dump_filename: Option<PathBuf>)
-    -> Result<YkExtraLinkObject, Box<dyn Error>> {
+    tcx: &'a TyCtxt<'a, 'tcx, 'gcx>, def_ids: &DefIdSet, mode: TirMode)
+    -> Result<Option<YkExtraLinkObject>, Box<dyn Error>>
+{
+    let tir_path = do_generate_tir(tcx, def_ids, &mode)?;
+    match mode {
+        TirMode::Default(_) => {
+            // In this case the file at `tir_path` is a raw binary file which we use to make an
+            // object file for linkage.
+            let obj = YkExtraLinkObject::new(&tir_path, SECTION_NAME);
+            fs::remove_file(tir_path)?; // Now we have our object, we can remove the temp file.
+            Ok(Some(obj))
+        },
+        TirMode::TextDump(_) => {
+            // In this case we have no object to link, and we keep the file at `tir_path` around,
+            // as this is the text dump the user asked for.
+            Ok(None)
+        }
+    }
+}
 
-    // The filename must be the same between builds for the reproducible build tests to pass.
-    let mut mir_path: String = exe_filename.to_str().unwrap().to_owned();
-    mir_path.push_str(TMP_EXT);
-    let mut fh = File::create(&mir_path)?;
-    let mut enc = ykpack::Encoder::from(&mut fh);
+fn do_generate_tir<'a, 'tcx, 'gcx>(
+    tcx: &'a TyCtxt<'a, 'tcx, 'gcx>, def_ids: &DefIdSet, mode: &TirMode)
+    -> Result<PathBuf, Box<dyn Error>>
+{
+    let (tir_path, mut default_file, textdump_file) = match mode {
+        TirMode::Default(exe_path) => {
+            // The default mode of operation dumps TIR in binary format to a temporary file, which
+            // is later converted into an ELF object. Note that the temporary file name must be the
+            // same between builds for the reproducible build tests to pass.
+            let mut tir_path = exe_path.clone();
+            tir_path.set_extension(TMP_EXT);
+            let mut file = File::create(&tir_path)?;
+            (tir_path, Some(file), None)
+        },
+        TirMode::TextDump(dump_path) => {
+            // In text dump mode we just write lines to a file and we don't need an encoder.
+            let mut file = File::create(&dump_path)?;
+            (dump_path.clone(), None, Some(file))
+        },
+    };
+
+    let mut enc = match default_file {
```

Comment:
`let mut enc = default_file.map(|ref mut f| Some(...));` or similar?

Reply:
Ah, yes.

Reply:
Hrm.

Seems the closure upsets the checker:
```
167 |     let mut enc = default_file.map(|ref mut f| ykpack::Encoder::from(f));
    |                                     ^^^^^^^^^                         - temporary value dropped here while still borrowed
    |                                     |
    |                                     temporary value does not live long enough
```

Any ideas?

Reply:
Although I doubt it works, does `|mut f|` fix things?

Reply:
Funnily enough, if I remove the `ref` it asks me to also remove `mut`, so the closest we would get is:

```
169 |     let mut enc = default_file.map(|f| ykpack::Encoder::from(&mut f));
    |                                                                   ^- `f` dropped here while still borrowed
    |                                                                   |
    |                                                                   borrowed value does not live long enough
...
206 | }
    | - borrowed value needs to live until here
```

Reply:
OK, let's give up ;)